metrics-for-spark-clusters.md

```diff
@@ -0,0 +1,248 @@
+feature should always be deployed. For this specification, we should assume
+that the metrics server is always deployed.
+
+The Spark images currently in use will need to have a `metrics.properties`
+file added to their configuration directory (`$SPARK_HOME/conf`). This file
+is responsible for enabling the metrics and the configuring the sink in use.
+The necessary components of that file can be seen in example 3.
+
+Images will need to be created for the Graphite API server and the Carbon
+storage application. These images should be based on Centos for public
+solutions and hosted on the Docker registry alongside current images. In this
+manner, these images can be used by any Oshinko deployment. As these images
+may require customization beyond a simple docker file, directories should be
+created in the oshinko-rest project under the top-level `tools` directory to
+house them.
+
+It is worth noting that one of the primary strengths of a Graphite-based
+solution is the ease of access to the underlying metrics data. The Graphite
+API server exposes a RESTful endpoint for querying the data, this interface
+allows retrieval of the data in several structured formats (eg. JSON) as well
+as graphical representations (eg. SVG).
+
+**Example 2, DeploymentConfig for Spark master with Graphite**
+
+```
+- kind: DeploymentConfig
+  apiVersion: v1
+  metadata:
+    name: ${MASTER_NAME}
+    labels:
+      oshinko-cluster: ${CLUSTER_NAME}
+      oshinko-type: master
+  spec:
+    strategy:
+      type: Rolling
+    triggers:
+      - type: ConfigChange
+    replicas: 1
+    selector:
+      name: ${MASTER_NAME}
+    template:
+      metadata:
+        labels:
+          name: ${MASTER_NAME}
+          oshinko-cluster: ${CLUSTER_NAME}
+          oshinko-type: master
+      spec:
+        volumes:
+          - name: whisper
+            emptyDir:
+              medium: ""
+        containers:
+          - name: ${MASTER_NAME}
+            image: ${SPARK_MASTER_IMAGE}
+            env:
+              - name: SPARK_MASTER_PORT
+                value: "7077"
+              - name: SPARK_MASTER_WEBUI_PORT
+                value: "8080"
+            ports:
+              - containerPort: 7077
+                protocol: TCP
+              - containerPort: 8080
+                protocol: TCP
+          - name: ${MASTER_NAME}-graphite
+            image: ${GRAPHITE_IMAGE}
+            ports:
+              - containerPort: 8000
+                protocol: TCP
+            volumeMounts:
+              - name: whisper
+                mountPath: /var/lib/carbon/whisper
+          - name: ${MASTER_NAME}-carbon
+            image: ${CARBON_IMAGE}
+            ports:
+              - containerPort: 2003
+                protocol: TCP
+            volumeMounts:
+              - name: whisper
+                mountPath: /var/lib/carbon/whisper
+```
+
+**Example 3, metrics.properties important bits**
+
+```
+*.sink.graphite.class=org.apache.spark.metrics.sink.GraphiteSink
```

Comment:
that IP address is just from my testing configs, when this is deployed to OpenShift it can use 127.0.0.1 for the IP address as all the containers can speak through localhost in a pod.

i don't think we'll need a dynamic update, so we *should* be able to build it into the image.